ls1intum/Artemis
# Fix combining template commits in Gitlab

Fixes #2434




### Checklist

### Motivation and Context



The "Combine Template Commits" functionality in the programming exercises does not work if Gitlab is used instead of Bitbucket even if the client notifies the user that it succeeded. The reason is that Gitlab protects the master branch by default, leading to an error when force-pushing into that branch.

The solution to this problem is to unprotect the template repository master branch  when creating/importing a programming exercise.

### Description

1. I extended the `VersionControlService` by two methods:
  - `void protectBranch(URL repositoryUrl, String branch)` (not used for added for the sake of completion) 
  - `void unprotectBranch(URL repositoryUrl, String branch)`
2. I implemented these method in the `GitlabService`. The implementation in `BitBucketService` is empty because it's not in use.
3. I unprotected the master branch of the template repository when a programming exercise is created/imported.

### Steps for Testing

Please use test server 2 for testing.

1. Log in to Artemis and create a new programming exercise. Make sure you select the "Online Editor" feature.
2. Open up Gitlab and go to the exercise repository of the programming exercise.
3. Click on Repository -> Branches and make sure the master branch is not protected (only "default" is displayed).
4. Click on Repository -> Commits. There should be a single commit.
5. Go back to the Course Administration -> Course -> programming exercise and open the online editor (as intructor)
6. Make a change in the template repository and submit it.
7. Make another change in the template repository and submit it.
3. Go back to the repository in Gitlab and check that there three commits now.
4. Go back to the Course Administration -> Course -> programming exercise -> View -> Combine Template Commits
5. Go back to the repository in Gitlab and check that there's one commit now.

**Note:** Normally this shouldn't affect the way students modify code.

### Test Coverage
Didn't change
